File: samples/dark_cyber.scala
/*
 * In a land shrouded in mystery ....
 */
class DarkCyber(name: String, constructionYear: Int, destructionYear: Int, features: Array[String]) {
  def duration(): Int = {
    calcDuration()
  }

  def calcDuration(): Int = {
    this.destructionYear - this.constructionYear
  }
}

// ...there stood a Dark Cyber
val moonshadowFortress = new DarkCyber(name = "Moonshadow Fortress", constructionYear = 980, destructionYear = 1503, features = Array("Eerie Glow", "High Battlements"))

println(s"The duration of ${moonshadowFortress.name} was ${moonshadowFortress.duration()} years.")

File: samples/dark_cyber.ts
/*
 * In a land shrouded in mystery ...
 */

interface DarkCyberProps {
  name: string;
  constructionYear: number;
  destructionYear: number;
  features: string[];
}

class DarkCyber {
  name: string;
  constructionYear: number;
  destructionYear: number;
  features: string[];

  constructor(props: DarkCyberProps) {
    this.name = props.name;
    this.constructionYear = props.constructionYear;
    this.destructionYear = props.destructionYear;
    this.features = props.features;
  }

  get duration(): number {
    return this.calcDuration();
  }

  calcDuration(): number {
    return this.destructionYear - this.constructionYear;
  }
}

// ...there stood a Dark Cyber

const MoonshadowFortress = new DarkCyber({
  name: "Moonshadow Fortress",
  constructionYear: 980,
  destructionYear: 1503,
  features: ["Eerie Glow", "High Battlements"],
});

console.log(`The duration of ${MoonshadowFortress.name} was ${MoonshadowFortress.duration} years.`);
